vochain/account.go

```diff
@@ -192,94 +171,406 @@ func (v *State) VerifyAccountBalance(message, signature []byte, amount uint64) (
+		return nil, fmt.Errorf("cannot extract public key from signature: %w", err)
 	}
-	// check account, if not exists a new one will be created
-	acc, err := state.GetAccount(accountAddress, false)
+	returnValues.TxSender, err = ethereum.AddrFromPublicKey(pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("cannot extract address from public key: %w", err)
+	}
+	// get txSender account
+	txSender, err := state.GetAccount(returnValues.TxSender, false)
+	if err != nil {
+		return nil, fmt.Errorf("cannot check if account %s exists: %v", returnValues.TxSender.String(), err)
+	}
+	// get tx.Account, if not valid set to txSender address
+	accountAddressBytes := tx.GetAccount()
+	if accountAddressBytes == nil ||
+		len(accountAddressBytes) != types.EntityIDsize ||
+		bytes.Equal(accountAddressBytes, types.EthereumZeroAddressBytes[:]) {
+		returnValues.Account = returnValues.TxSender
+	} else {
+		returnValues.Account = common.BytesToAddress(accountAddressBytes)
+	}
+	// if not exist create new one
+	if txSender == nil {
+		if tx.FaucetPackage != nil {
+			if tx.FaucetPackage.Payload != nil {
+				if bytes.Equal(tx.FaucetPackage.Payload.To, returnValues.TxSender.Bytes()) {
+					// get issuer address from faucetPayload
+					faucetPkgPayload := tx.FaucetPackage.GetPayload()
+					faucetPackageBytes, err := proto.Marshal(faucetPkgPayload)
+					if err != nil {
+						return nil, fmt.Errorf("cannot extract faucet package payload: %v", err)
+					}
+					issuerAddress, err := ethereum.AddrFromSignature(faucetPackageBytes, tx.FaucetPackage.Signature)
+					if err != nil {
+						return nil, err
+					}
+					// check issuer nonce not used
+					b := make([]byte, 8)
+					binary.LittleEndian.PutUint64(b, faucetPkgPayload.Identifier)
+					key := issuerAddress.Bytes()
+					key = append(key, b...)
+					used, err := state.FaucetNonce(crypto.Sha256(key), false)
+					if err != nil {
+						return nil, fmt.Errorf("cannot check faucet nonce: %v", err)
+					}
+					if used {
+						return nil, fmt.Errorf("nonce %d already used", faucetPkgPayload.Identifier)
+					}
+					// check issuer have enough funds
+					issuerAcc, err := state.GetAccount(issuerAddress, false)
+					if err != nil {
+						return nil, fmt.Errorf("cannot get faucet account: %v", err)
+					}
+					if issuerAcc == nil {
+						return nil, fmt.Errorf("the account signing the faucet payload does not exist")
+					}
+					cost, err := state.TxCost(models.TxType_COLLECT_FAUCET, false)
+					if err != nil {
+						return nil, fmt.Errorf("cannot get %s tx cost: %w", models.TxType_COLLECT_FAUCET, err)
+					}
+					if (issuerAcc.Balance) < faucetPkgPayload.Amount+cost {
+						return nil, fmt.Errorf("faucet does not have enough balance %d < %d", issuerAcc.Balance, faucetPkgPayload.Amount+cost)
+					}
+					returnValues.FaucetPayloadSigner = issuerAddress
+					returnValues.FaucetPayload = tx.GetFaucetPackage().Payload
+				} else {
+					return nil, fmt.Errorf("payload to and tx sender missmatch")
```

Comment:
tip: if you reverse this if and return early, then you can unindent the bulk of the code. same for some other "ifs" that have this pattern.